nixos/nextcloud: add extraOptions and secretFile options

stuebinm opened the pull request:
###### Motivation for this change

Issue: Fixes #116189

This adds two options to the nextcloud module, with the aim of allowing the entirety of `config.php` to be set declaratively (while allowing the user to choose for which options this makes sense, since nextcloud also keeps state in there):
1. `services.nextcloud.extraOptions`, which takes free-form options given as an attribute set, and reads them in via json from the php side (which lets us prevent syntax errors in php, if not key errors – given the full length of nextcloud's potential options, I don't think specifying them all via Nix is a viable option)
2. `services.nextcloud.secretFile`, which takes a path to a json file specifying options which a user may want to keep secret and not add to their nix store; this file is read in the same way on startup by php, and may even overwrite options set in some other way.

Additionally, adds a new test for nextcloud based on the current with-postgresql-and-redis test, but with a more declarative config instead of using `nextcloud-occ`.

###### Things done
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

stuebinm committed:
nixos/nextcloud: allow more declarative config

This adds two options to the nextcloud module, with the aim of allowing
the entirety of `config.php` to be set declaratively:
1. `services.nextcloud.extraOptions`, which takes free-form options
   given as an attribute set, and reads them in via json from the php
   side (which lets us prevent syntax errors in php, if not key errors –
   given the full length of nextcloud's potential options, I don't think
   specifying them all via Nix is a viable option)
2. `services.nextcloud.secretFile`, which takes a path to a json file
   specifying options which a user may want to keep secret and not add
   to their nix store; this file is read in the same way on startup by
   php, and may even overwrite options set in some other way.

stuebinm committed:
nixos/nextcloud: add test for declaratively defined redis

This is based on the test using redis and postgresql, but it does not
require any imperative configuration after startup; everything is defined
via Nix instead.

lheckemann commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -480,6 +519,11 @@ in {
               ${optionalString (c.defaultPhoneRegion != null) "'default_phone_region' => '${c.defaultPhoneRegion}',"}
             ];
+
+            $EXTRACONFIG = json_decode('${builtins.toJSON cfg.extraOptions}', true);
```
These should probably be read from a file (though writing it to the store is fine here) as well, as the JSON may contain single-quote characters.

stuebinm replied:
implemented that it https://github.com/NixOS/nixpkgs/pull/118093/commits/83a669a0bef463b3202a544b33efc9109605ce83

stuebinm replied:
Could you specify what you mean with RFC42-style approach? As far as I can see it suggests doing these extra-options like hooks as structured attribute sets instead of strings, which is what I have done (via json formatting, since I couldn't find any better way to serialise to php).

Apart from that, the rfc talks a lot about balancing these options with "normal" nixos module options, but I didn't really want to replace any of the other parts, since the whole module is quiet complex and I don't actually understand any php (I just needed a way for secrets & extra options in my own deployment). I guess we could rename `extraOptions` to `extraSettings` though.

lheckemann reviewed (commented):
:+1: thanks! I'm always happy to see more modules supporting keeping secrets out of the store :)

For the `extraOptions` part, I think an [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md)-style approach would be nicer, since it integrates better with the module system, though I'd defer to people who are more involved with this module on that topic. cc @ma27 :)

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -480,6 +519,11 @@ in {
+
+            $EXTRACONFIG = json_decode('${builtins.toJSON cfg.extraOptions}', true);
+
+            array_push($CONFIG, $EXTRACONFIG);
```
Is `array_push` the thing you actually need here?

```
>>> $foo=[1,2,3,4];
=> [
     1,
     2,
     3,
     4,
   ]
>>> $b=[4,5,6,7];
=> [
     4,
     5,
     6,
     7,
   ]
>>> array_push($foo,$b);
=> 5
>>> $foo
=> [
     1,
     2,
     3,
     4,
     [
       4,
       5,
       6,
       7,
     ],
   ]

```

stuebinm replied:
resolved in a different way (see your other thread further down)

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -367,6 +367,32 @@ in {
+      description = ''
+        Extra options which should be appended to nextcloud's config.php file.
+      '';
+      example = {
```
IIRC using `literalExample` makes better formatting in the manual.

Ma27 replied:
It's correct though becaues it gives you better formatting.

Ma27 replied:
Please use `literalExample '' /* nix value */ ''` here.

stuebinm replied:
done

stuebinm committed:
nixos/nextcloud: use array_merge instead of array_push

This should correctly merge the various configs (via nix module options,
the secretFile option, and the extraOptions option).

stuebinm committed:
nixos/nextcloud: better json typechecking

this now uses the pkgs.formats.json instead of the builtin function to
serialise json, which comes with a type signature that we can pass to mkOptions

stuebinm commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -367,6 +367,32 @@ in {
     };
+
+    extraOptions = mkOption {
+      type = with types; attrsOf anything;
```
I've replaced this with the type of pkgs.formats.json, so it should be fine now.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -367,6 +369,33 @@ in {
+      description = ''
+        Secret options which will be appended to nextcloud's config.php file (written as JSON, in the same
+        form as the <xref linkend="opt-services.nextcloud.extraOptions"/> option), for example
+        '{"redis":{"password":"secret","timeout"=2}}'.
```
Tbh I don't really like that interface. Can you at least make sure that one can write
```
{
  "foo": "bar"
}
```

Also, I'm wondering if we want to do some kind of `file_get_contents`-logic for a single option specifically. And I don't really understand why `timeout` has to be here.

stuebinm replied:
this should already work.

The "timeout"=2 bit was a mistake (it don't think it should hurt to set it here, though using a "=" for that is definitely wrong since it's not valid json); I've removed it.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -457,6 +486,19 @@ in {
               }
             ''}
+            ${optionalString (cfg.secretFile != null) ''
+              function nix_read_secrets() {
```
This should be deduplicated IMHO. A `nix_read_file($filename)` could be made for instance.

stuebinm replied:
done (with an additional parameter $error to model error messages in case the file could not be read)

stuebinm replied:
note that there's now two functions: the nix_read_file function suggested here, and a nix_decode_json_file function which does the same but also handles json decoding & potential errors

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -480,6 +522,11 @@ in {
               ${optionalString (c.defaultPhoneRegion != null) "'default_phone_region' => '${c.defaultPhoneRegion}',"}
             ];
+
+            $EXTRACONFIG = json_decode(file_get_contents("${jsonFormat.generate "nextcloud-extraOptions.json" cfg.extraOptions}"), true);
```
This creates a store-path for no reason. Any reason why we don't use `builtins.toJSON` here? Also, `json_decode` will silently fail if something goes wrong with the decode. You may want to check with `json_last_error() !== 0` and throw an error with `json_last_error_msg()` in the message.

stuebinm replied:
I recall the reason for this was that I couldn't figure out how to escape the json string produced by `builtins.toJSON` in a way that wouldn't crash php, since the only way to write a string in php appears to be quoting it with " which json also uses. Presumably the string could be escaped, but since I'm unfamiliar with the subtleties of php strings it seemed safer to do it this way.

(As I think I mentioned previously, I don't actually know much about php; the original motivation for this PR is that I needed a secretFile option and hacked one in – so while this should work, the php code may not necessarily be very elegant or idiomatic)

Ma27 replied:
Ahh you're right! While I'm not happy with this, this would escalate way to fast (I checked a few ideas, but I couldn't find a solution that's worth the effort).
However, it'd still be nice to have a safety guard here regarding how to behave if the JSON cannot be decoded by PHP. This could by implemented like this (right after the declaration of `$EXTRACONFIG`):

``` php
if (json_last_error() !== JSON_ERROR_NONE) {
  throw new \RuntimeException(sprintf("Cannot decode %s, because: %s", $filename, json_last_error_msg()));
}
```

stuebinm replied:
done (though in the nix_decode_json_file function)

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -480,6 +522,11 @@ in {
+            $EXTRACONFIG = json_decode(file_get_contents("${jsonFormat.generate "nextcloud-extraOptions.json" cfg.extraOptions}"), true);
+
+            $CONFIG = array_merge($CONFIG, $EXTRACONFIG);
+            ${optionalString (cfg.secretFile != null) "$CONFIG = array_merge($CONFIG, nix_read_secrets());"}
```
`array_replace_recursive()` may be better here though.

Ma27 commented on nixos/tests/nextcloud/with-declarative-redis.nix:
```diff
@@ -0,0 +1,103 @@
+import ../make-test-python.nix ({ pkgs, ...}: let
+  adminpass = "hunter2";
```
I like that :D

Ma27 commented on nixos/tests/nextcloud/with-declarative-redis.nix:
```diff
@@ -0,0 +1,103 @@
+    client.succeed(
+        "${withRcloneEnv} ${diffSharedFile}"
+    )
+  '';
```
I think it'd make more sense here to actually test if the redis-cache itself is working.

stuebinm replied:
I agree. However, I am unsure how to do that, and since I copied the entire test from `with-postgresql-and-redis.nix` and just changed the redis configuration, I though the test should be sufficient.

Ma27 replied:
Perhaps try to upload a file to nextcloud with `rclone` (the `basic.nix` shows how to do that) and then check if something was written to the cache (no idea if that works, but such a test IMHO makes sense to check for potential regression in the cache config in the future).

Ma27 reviewed (commented):
Sorry, I forgot about this. Left a few more comments.

Ma27 commented:
@stuebinm may I ask if you're still interested in working on this? :)

stuebinm committed:
nixos/nextcloud: deduplicate file reading

This combines the two functions nix_read_pwd and nix_read_secrets into a
single function nix_read_file, which takes two arguments: the file to be
read, and an error message that should be printed if reading the file
failed.

stuebinm committed:
nixos/nextcloud: test for secretFile option

This is a dummy test that doesn't do much — the secretFile option is
just used to set the database type to postgres; otherwise this is an
exact copy of the with-postgresql-and-redis.nix test, though with the
redis components removed.

stuebinm committed:
nixos/nextcloud: fixed secretFile example

stuebinm commented:
> @stuebinm may I ask if you're still interested in working on this? :)

Huh. I replied to you a week ago, but apparently github's mailin just quietly failed on us:

> I am, but it's also not an immediate concern for me and in any case I won't have time for it until at least the next week (and possibly not for the next three weeks, depending on how other things turn out). Though tbh I also almost forgot this MR was even still open …

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-add-custom-options-to-the-deeply-nested-nextcloud-configuration-file/14567/1

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -367,6 +369,33 @@ in {
+      description = ''
+        Secret options which will be appended to nextcloud's config.php file (written as JSON, in the same
+        form as the <xref linkend="opt-services.nextcloud.extraOptions"/> option), for example
+        '{"redis":{"password":"secret"}}'.
```
(nit for formatting) please use e.g. `<programlisting>` here for better readability (there should be enough examples in the manual already :))

stuebinm replied:
both done

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -474,12 +496,30 @@ in {
+
+            $CONFIG = array_merge($CONFIG, $EXTRACONFIG);
+            ${optionalString (cfg.secretFile != null) ''
+              $CONFIG = array_merge($CONFIG, nix_read_file(
```
I'd say that in both cases, `array_replace_recursive` is the better choice here.

Ma27 reviewed (commented):
> Huh. I replied to you a week ago, but apparently github's mailin just quietly failed on us:

@stuebinm That's funny, the same thing happened to me, only learnt that here are updates because of yesterday's discourse notification :D

Ma27 commented:
@stuebinm do you intend to continue working on this? :)

Ma27 commented:
Still relevant.

stuebinm committed:
services/nextcloud: apply suggestions from PR 118093

stuebinm committed:
services/nixos: decode secret file correctly

stuebinm committed:
services/nextcloud: check redis config

stuebinm committed:
services/nextcloud: better secretFile test

this is virtually identical to the declarative redis test now, except
that the config json is set manually.

stuebinm committed:
Merge branch 'master' into nextcloud-secrets

stuebinm commented:
anyone want to take a look at this?

Ma27 commented:
@stuebinm how much is still missing here in you opinion to get it merged? It would help if you could resolve the merge conflicts and resolve all of my review comments that you consider fixed.

I don't really have the time to start yet another side-project (i.e. implementing missing stuff), but if it's just a review away from a merge after that, I'm happy to do that ;-)

jtrees commented:
Landed here after failing to find `extraOptions`. Very much interested.

kevincox commented:
The module doesn't have any maintainers but let's bring in the package maintainers. I like this approach, structured configuration is always better than bashing strings together. Once this is merged we can convert the other options to this format.

@schneefux @bachp @globin

@stuebinm if you could update the PR to fix conflicts in the meantime that would be great. If the maintainers don't speak up I think we can merge this.

Ma27 commented:
> The module doesn't have any maintainers but let's bring in the package maintainers

Uhm, what about me? %) 

I'll try to be able to take a look soonish.

kevincox commented:
There is no metadata listing maintainers on the module. But that is a newer thing IIRC so I'm not surprised it may be missing.

Thanks for taking a look.

kevincox commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -684,13 +723,34 @@ in {
+                "${cfg.secretFile}",
+                "Cannot start Nextcloud, secrets file %s set by NixOS doesn't exist!"
+              ));
+            ''}
```
Nit: Why are these different. Can we just do the one-line version for both extraconfig and secret?

stuebinm replied:
Indeed we can! 

I think the reason why it's like this was mostly historical — this PR is pretty old and saw a lot of changes

stuebinm committed:
services/nextcloud: more consistent code

stuebinm committed:
Merge branch 'master' into nextcloud-secrets

stuebinm commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -728,14 +767,34 @@ in {
+
+            $CONFIG = array_replace_recursive($CONFIG, nix_decode_json_file(
+              "${jsonFormat.generate "nextcloud-extraOptions.json" cfg.extraOptions}",
+              "decoding json extra options file failed"
```
maybe?

I guess we could do this:

```nix
${ let extraConfig = jsonFormat.generate "nextcloud-extraOptions.json" cfg.extraOptions;
   in ''
     $CONFIG = array_replace_recursive($CONFIG, nix_decode_json_file(
     "${extraConfig}",
     "decoding the json-encoded file generated from the extraOptions failed (${extraConfig}). This is probably a bug in nixpkgs."
   ));
 ''}
```

stuebinm replied:
ah — tbh i don't like writing php (never properly learnt it, and every time i touch anything things seem to go up in flames), so i tend to do as much as possible on the nix side of things. But if you write a snippet / patch for that I'll be happy to add it in

Ma27 replied:
> This doesn't have %s although it can theoretically never fail. Maybe it makes more sense to just pass in "file type" such as "extraOptions file" or "secretFile file" and the function itself can generate a message and include the path?

AFAIU this is nothing an admin can break while deploying bogus configuration, but only if someone contributes code that's wrong. So I'd argue it's good enough and other failures just shouldn't pass reviews.

kevincox replied:
I don't like passing in an invalid format string because it can't be used. Let's either pass in something trivial like `"this should never happen %s"` or handle this properly. Honestly i don't think we need to do anything too fancy, but I don't want to merge in "wrong" code as it will confuse future readers.

stuebinm replied:
I've changed it

stuebinm commented:
I've resolved the threads which seemed either outdated or resolved to me @Ma27

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -547,6 +549,33 @@ in {
+      description = ''
+        Extra options which should be appended to nextcloud's config.php file.
+      '';
+      example = literalExample '' {
```
Should be `literalExpression` IIRC.

stuebinm replied:
changed it

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -728,14 +767,34 @@ in {
               ${optionalString (c.dbtableprefix != null) "'dbtableprefix' => '${toString c.dbtableprefix}',"}
-              ${optionalString (c.dbpassFile != null) "'dbpassword' => nix_read_secret('${c.dbpassFile}'),"}
+              ${optionalString (c.dbpassFile != null) ''
+                  dbpassword' => nix_read_file(
```
Shouldn't this be `nix_read_secret`?

stuebinm replied:
it should be, yes! seems to have been forgotten at some point in this PR's history …

Ma27 reviewed (commented):
Did a review, overall LGTM, just added a few minor notes.
@kevincox would you mind testing the change before merging, not sure if I'll get to it soonish :)

kevincox commented:
I can probably test but it doesn't look ready yet. We have some function name errors so IIUC this doesn't evaluate.

stuebinm committed:
services/nextcloud: fix a bug

stuebinm committed:
services/nextcloud: apply suggestions

stuebinm commented on nixos/tests/nextcloud/default.nix:
```diff
@@ -16,6 +16,14 @@ foldl
+      inherit system pkgs;
+      nextcloudVersion = ver;
+    };
+    "with-secrets${toString ver}" = import ./with-secrets.nix {
```
done that too

stuebinm commented:
I've looked at it again, the nextcloud tests all run through fine for me, and it's reasonably up to date with the master branch. Also I've removed a space on an empty line somewhere that made one of the checks fail.

kevincox commented:
Result of `nixpkgs-review pr 118093` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

kevincox commented:
I updated my server to this and used it to configure my SMTP settings. It worked like a charm.

stuebinm committed:
services/nextcloud: impossible error message

kevincox reviewed (approved):
Looks good to me. @Ma27 good to merge?

Ma27 commented:
Oh sorry forgot to reply, sgtm :)

kevincox commented:
I figured we can fix any last things in followups :)
